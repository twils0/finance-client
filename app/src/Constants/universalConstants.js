import stripeConfig from '../../../stripe.config.json';
import dbConfig from '../../../db.config.json';
import iexConfig from '../../../iex.config.json';

export const baseURLs = {
  STRIPE: stripeConfig.url,
  DB: dbConfig.url,
  IEX: iexConfig.urls.api,
  IEX_DEV: iexConfig.urls.dev,
  IEX_TERMS: iexConfig.urls.terms,
};

export const URLs = {
  CUSTOMERS: '/customers',
  USERS: '/users',
  EMAILS: '/emails',
  SECURITIES: '/securities',
  SECURITIES_SEARCH: '/securities/search',
  IEX_COMPANY: tickerCusip => `/stock/${tickerCusip}/company`,
  IEX_QUOTE: tickerCusip => `/stock/${tickerCusip}/quote`,
};

export const axiosConfig = {
  STRIPE: {
    baseURL: baseURLs.STRIPE,
    timeout: 30000,
  },
  DB: {
    baseURL: baseURLs.DB,
    timeout: 30000,
  },
  IEX: {
    baseURL: baseURLs.IEX,
    timeout: 30000,
  },
};

export const pathNames = {
  DEFAULT: '*',
  LOGIN: '/login',
  SIGN_UP: '/signup',
  CODE: '/code',
  CODE_VERIFY_EMAIL: '/code/:verificationId?',
  TERMS: '/terms',
  INTERNAL: '/internal',
  WATCHLIST: '/internal/watchlist',
  WATCHLIST_SECURITY_ID: '/internal/watchlist/:securityId?',
  ACCOUNT: '/internal/account',
};

export const requestStatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
};
