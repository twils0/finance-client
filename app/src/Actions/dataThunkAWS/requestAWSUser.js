import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestLogout from '../dataThunkAuth/requestLogout';

import { demo } from '../../../../mode.config.json';

const requestAWSUser = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.aws.status !== requestStatusTypes.LOADING) {
    dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

    let user = null;

    try {
      if (!demo) {
        user = await Auth.currentUserPoolUser();
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));

      if (
        (typeof error === 'string'
          && (error === 'not authenticated' || error === 'No current user in userPool'))
        || (typeof error === 'object' && error && error.code === 'NotAuthorizedException')
      ) {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'requestAWSUser',
        });
      }

      return Promise.reject(error);
    }

    if (!demo) {
      dispatch(setAWSUser({ user }));
    }
    dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));

    return { user };
  }

  return null;
};

export default requestAWSUser;
