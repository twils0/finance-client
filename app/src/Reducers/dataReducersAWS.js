import { combineReducers } from 'redux';

import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes } from '../Constants/dataConstantsAWS';

export const initialState = {
  status: requestStatusTypes.IDLE,
  user: null,
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_STATUS:
      return action.payload.status;
    case actionTypes.RESET_AWS_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const user = (state = initialState.user, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_USER:
      return action.payload.user;
    case actionTypes.RESET_AWS_STATE:
      return initialState.user;
    default:
      return state;
  }
};

const dataReducersAWS = combineReducers({
  status,
  user,
});

export default dataReducersAWS;
