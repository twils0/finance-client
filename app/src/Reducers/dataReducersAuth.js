import { combineReducers } from 'redux';

import { requestStatusTypes, pathNames } from '../Constants/universalConstants';
import { actionTypes, statusNames, codeTypeNames } from '../Constants/dataConstantsAuth';

export const initialState = {
  authenticated: false,
  redirectURL: pathNames.WATCHLIST,
  status: {
    [statusNames.LOGIN]: {
      id: statusNames.LOGIN,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_OUT_DEVICES]: {
      id: statusNames.SIGN_OUT_DEVICES,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGIN_MFA]: {
      id: statusNames.LOGIN_MFA,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_UP]: {
      id: statusNames.SIGN_UP,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE]: {
      id: statusNames.VERIFY_PHONE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE_CODE]: {
      id: statusNames.VERIFY_PHONE_CODE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL]: {
      id: statusNames.VERIFY_EMAIL,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL_LINK]: {
      id: statusNames.VERIFY_EMAIL_LINK,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.FORGOT_PASSWORD]: {
      id: statusNames.FORGOT_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.RESET_PASSWORD]: {
      id: statusNames.RESET_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.CHANGE_PASSWORD]: {
      id: statusNames.CHANGE_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGOUT]: {
      id: statusNames.LOGOUT,
      status: requestStatusTypes.SUCCESS,
    },
    [statusNames.DELETE_ACCOUNT]: {
      id: statusNames.DELETE_ACCOUNT,
      status: requestStatusTypes.SUCCESS,
    },
  },
  codeTypes: {
    [codeTypeNames.VERIFY_PHONE]: {
      id: codeTypeNames.VERIFY_PHONE,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL]: {
      id: codeTypeNames.VERIFY_EMAIL,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL_ADDITIONAL]: {
      id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL,
      needed: true,
    },
  },
};

export const authenticated = (state = initialState.authenticated, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTHENTICATED:
      return action.payload.authenticated;
    case actionTypes.RESET_AUTH_STATE:
      return initialState.authenticated;
    default:
      return state;
  }
};

export const redirectURL = (state = initialState.redirectURL, action) => {
  switch (action.type) {
    case actionTypes.SET_REDIRECT_URL:
      return action.payload.redirectURL;
    default:
      return state;
  }
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTH_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const codeTypes = (state = initialState.codeTypes, action) => {
  switch (action.type) {
    case actionTypes.SET_CODE_TYPE:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.codeTypes;
    default:
      return state;
  }
};

const dataReducersAuth = combineReducers({
  authenticated,
  redirectURL,
  status,
  codeTypes,
});

export default dataReducersAuth;
