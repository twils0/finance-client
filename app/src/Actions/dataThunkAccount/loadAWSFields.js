import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import { requestStatusTypes } from '../../Constants/universalConstants';
import { codeTypeNames } from '../../Constants/dataConstantsAuth';
import { statusNames, fieldNames } from '../../Constants/dataConstantsAccount';
import { setCodeType } from '../dataActionsAuth';
import requestLogout from '../dataThunkAuth/requestLogout';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import { setAccountStatus, setFields } from '../dataActionsAccount';

import handleErrorCatch from '../../handleErrorCatch';

import { demo } from '../../../../mode.config.json';
import demoFields from '../../../../demo/demoFields';

const loadAWSFields = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.account.status[statusNames.AWS_FIELDS].status !== requestStatusTypes.LOADING) {
    dispatch(
      setAccountStatus({
        id: statusNames.AWS_FIELDS,
        status: requestStatusTypes.LOADING,
      }),
    );

    let { user } = state.data.aws;
    const payload = {};

    try {
      if (!user) {
        ({ user } = await dispatch(requestAWSUser()));
      }

      let fields = null;

      if (!demo) {
        fields = await Auth.userAttributes(user);
      } else {
        fields = demoFields;
      }

      fields.forEach((field) => {
        switch (field.Name) {
          case 'name':
            payload[fieldNames.NAME] = {
              id: fieldNames.NAME,
              value: field.Value,
            };
            break;
          case 'email':
            payload[fieldNames.EMAIL] = {
              id: fieldNames.EMAIL,
              value: field.Value,
            };
            break;
          case 'custom:email_ver': {
            dispatch(
              setCodeType({ id: codeTypeNames.VERIFY_EMAIL, needed: field.Value !== 'true' }),
            );
            break;
          }
          case 'custom:email_additional':
            payload[fieldNames.EMAIL_ADDITIONAL] = {
              id: fieldNames.EMAIL_ADDITIONAL,
              value: field.Value,
            };
            break;
          case 'custom:email_additional_ver':
            dispatch(
              setCodeType({
                id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL,
                needed: field.Value !== 'true',
              }),
            );
            break;
          case 'phone_number': {
            payload[fieldNames.PHONE] = {
              id: fieldNames.PHONE,
              value: field.Value,
            };
            break;
          }
          default:
            break;
        }
      });
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(
        setAccountStatus({
          id: statusNames.AWS_FIELDS,
          status: requestStatusTypes.ERROR,
        }),
      );

      if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'loadAWSFields',
        });
      }

      return Promise.reject(error);
    }

    dispatch(setFields(payload));

    dispatch(
      setAccountStatus({
        id: statusNames.AWS_FIELDS,
        status: requestStatusTypes.SUCCESS,
      }),
    );
  }

  return null;
};

export default loadAWSFields;
