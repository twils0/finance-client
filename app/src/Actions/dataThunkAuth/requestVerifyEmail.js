import axios from 'axios';
import raven from 'raven-js';

import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import requestLogout from './requestLogout';
import { setAuthStatus } from '../dataActionsAuth';

import handleErrorCatch from '../../handleErrorCatch';

import { demo } from '../../../../mode.config.json';

const requestVerifyEmail = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.auth.status[statusNames.VERIFY_EMAIL].status !== requestStatusTypes.LOADING) {
    dispatch(
      setAuthStatus({
        id: statusNames.VERIFY_EMAIL,
        status: requestStatusTypes.LOADING,
      }),
    );

    let { user } = state.data.aws;

    try {
      if (!user) {
        ({ user } = await dispatch(requestAWSUser()));
      }

      if (!demo) {
        const idToken = user.signInUserSession.idToken.jwtToken;
        const accessToken = user.signInUserSession.accessToken.jwtToken;

        await axios.post(
          URLs.EMAILS,
          {},
          {
            headers: {
              Authorization: idToken,
            },
            params: {
              accessToken,
            },
            ...axiosConfig.DB,
          },
        );
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(
        setAuthStatus({
          id: statusNames.VERIFY_EMAIL,
          status: requestStatusTypes.ERROR,
        }),
      );

      if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'requestVerifyEmail',
        });
      }

      return Promise.reject(error);
    }

    dispatch(
      setAuthStatus({
        id: statusNames.VERIFY_EMAIL,
        status: requestStatusTypes.SUCCESS,
      }),
    );
  }

  return null;
};

export default requestVerifyEmail;
