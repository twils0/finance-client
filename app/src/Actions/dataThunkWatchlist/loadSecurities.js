import axios from 'axios';
import raven from 'raven-js';

import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsWatchlist';
import { setWatchlistStatus, setSecuritiesAll } from '../dataActionsWatchlist';
import requestLogout from '../dataThunkAuth/requestLogout';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';

import handleErrorCatch from '../../handleErrorCatch';

const loadSecurities = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.watchlist.status[statusNames.SECURITIES].status !== requestStatusTypes.LOADING) {
    dispatch(
      setWatchlistStatus({
        id: statusNames.SECURITIES,
        status: requestStatusTypes.LOADING,
      }),
    );

    let { user } = state.data.aws;
    let securities = null;

    try {
      if (!user) {
        ({ user } = await dispatch(requestAWSUser()));
      }

      const idToken = user.signInUserSession.idToken.jwtToken;
      const accessToken = user.signInUserSession.accessToken.jwtToken;

      const response = await axios.get(URLs.USERS, {
        headers: {
          Authorization: idToken,
        },
        params: {
          accessToken,
        },
        ...axiosConfig.DB,
      });

      ({ securities } = response.data.body);
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(
        setWatchlistStatus({
          id: statusNames.SECURITIES,
          status: requestStatusTypes.ERROR,
        }),
      );

      if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
        dispatch(requestLogout());

        return null;
      }

      raven.captureException(error, {
        logger: 'loadSecurities',
      });

      return Promise.reject(error);
    }

    if (securities.list.length > 0) {
      dispatch(setSecuritiesAll({ securities }));
    }

    dispatch(
      setWatchlistStatus({
        id: statusNames.SECURITIES,
        status: requestStatusTypes.SUCCESS,
      }),
    );
  }

  return null;
};

export default loadSecurities;
