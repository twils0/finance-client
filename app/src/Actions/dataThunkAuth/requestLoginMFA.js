import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames, codeTypeNames } from '../../Constants/dataConstantsAuth';
import { setAuthStatus } from '../dataActionsAuth';
import requestVerifyEmail from './requestVerifyEmail';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import loadAWSFields from '../dataThunkAccount/loadAWSFields';
import loadSecurities from '../dataThunkWatchlist/loadSecurities';

const requestLoginMFA = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'code')) {
    throw new Error(`Please enter a value for the 'code' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    let state = getState();

    if (state.data.auth.status[statusNames.LOGIN_MFA].status !== requestStatusTypes.LOADING) {
      dispatch(
        setAuthStatus({
          id: statusNames.LOGIN_MFA,
          status: requestStatusTypes.LOADING,
        }),
      );

      let { user } = state.data.aws;

      try {
        if (!user) {
          ({ user } = await dispatch(requestAWSUser()));
        }

        await Auth.confirmSignIn(user, payload.code);

        dispatch(loadSecurities());

        await dispatch(loadAWSFields());

        state = getState();

        if (
          state.data.auth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed
          || state.data.auth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed
        ) {
          await dispatch(requestVerifyEmail());
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);

        dispatch(
          setAuthStatus({
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.ERROR,
          }),
        );

        if (
          typeof error === 'object'
          && error
          && (error.code === 'LimitExceededException'
            || error.code === 'ExpiredCodeException'
            || error.code === 'CodeMismatchException')
        ) {
          return Promise.reject(error);
        }

        raven.captureException(error, {
          logger: 'requestLoginMFA',
        });

        return Promise.reject(error);
      }

      dispatch(
        setAuthStatus({
          id: statusNames.LOGIN_MFA,
          status: requestStatusTypes.SUCCESS,
        }),
      );
    }
    return null;
  };
};

export default requestLoginMFA;
