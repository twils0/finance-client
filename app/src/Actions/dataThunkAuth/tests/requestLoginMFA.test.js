import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Auth } from 'aws-amplify';

import { initialState as initialStateAWS } from '../../../Reducers/dataReducersAWS';
import { initialState as initialStateAuth } from '../../../Reducers/dataReducersAuth';
import { requestStatusTypes } from '../../../Constants/universalConstants';
import {
  actionTypes as actionTypesAuth,
  codeTypeNames,
  statusNames,
} from '../../../Constants/dataConstantsAuth';
import requestLoginMFA from '../requestLoginMFA';
import requestVerifyEmail from '../requestVerifyEmail';
import requestAWSUser from '../../dataThunkAWS/requestAWSUser';
import loadAWSFields from '../../dataThunkAccount/loadAWSFields';
import loadSecurities from '../../dataThunkWatchlist/loadSecurities';

jest.mock('aws-amplify', () => ({
  Auth: {
    confirmSignIn: jest.fn(),
  },
}));
jest.mock('../../dataThunkAWS/requestAWSUser', () => jest.fn());
jest.mock('../../dataThunkAccount/loadAWSFields', () => jest.fn());
jest.mock('../../dataThunkWatchlist/loadSecurities', () => jest.fn());
jest.mock('../requestVerifyEmail', () => jest.fn());

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const userObject = { user: { test: 'testUser' } };
const mfaPayload = { code: '123456' };

describe('dataThunkAuth', () => {
  describe('requestLoginMFA', () => {
    afterEach(() => {
      requestAWSUser.mockReset();
      Auth.confirmSignIn.mockReset();
      loadAWSFields.mockReset();
      loadSecurities.mockReset();
      requestVerifyEmail.mockReset();
    });

    it("fails and throws an error when missing 'code' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      Auth.confirmSignIn.mockReturnValue(Promise.resolve());

      const emptyPayload = {};

      try {
        await store.dispatch(requestLoginMFA(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).not.toBeCalled();
      expect(loadAWSFields).not.toBeCalled();
      expect(requestVerifyEmail).not.toBeCalled();
    });

    it("calls requestAWSUser when missing 'user' object in state", async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;

      const store = mockStore({ data: { aws: initialStateAWS, auth: stateBeforeAuth } });

      requestAWSUser.mockReturnValue(() => Promise.resolve(userObject));
      Auth.confirmSignIn.mockReturnValue(Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());

      await store.dispatch(requestLoginMFA(mfaPayload));

      expect(requestAWSUser).toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).toBeCalled();
      expect(loadSecurities).toBeCalled();
      expect(requestVerifyEmail).not.toBeCalled();
    });

    it('fails and returns promise reject when requestAWSUser throws an error', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;

      const store = mockStore({ data: { aws: initialStateAWS, auth: stateBeforeAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      requestAWSUser.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestLoginMFA(mfaPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).toBeCalled();
      expect(Auth.confirmSignIn).not.toBeCalled();
      expect(loadAWSFields).not.toBeCalled();
      expect(loadSecurities).not.toBeCalled();
      expect(requestVerifyEmail).not.toBeCalled();
    });

    it('fails and returns promise reject when Auth.confirmSignIn throws an error', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: stateBeforeAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.confirmSignIn.mockReturnValue(Promise.reject(error));

      try {
        await store.dispatch(requestLoginMFA(mfaPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).not.toBeCalled();
      expect(loadSecurities).not.toBeCalled();
      expect(requestVerifyEmail).not.toBeCalled();
    });

    it('fails and returns promise reject when loadAWSFields throws an error', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: stateBeforeAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.confirmSignIn.mockReturnValue(Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestLoginMFA(mfaPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).toBeCalled();
      expect(loadSecurities).toBeCalled();
      expect(requestVerifyEmail).not.toBeCalled();
    });

    it('fails and returns promise reject when requestVerifyEmail throws an error, when email verify and email additional verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = true;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = true;
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: stateBeforeAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.confirmSignIn.mockReturnValue(Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      requestVerifyEmail.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestLoginMFA(mfaPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).toBeCalled();
      expect(loadSecurities).toBeCalled();
      expect(requestVerifyEmail).toBeCalled();
    });

    it('creates the correct actions with the correct payloads, when email additional verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = true;
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: stateBeforeAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.confirmSignIn.mockReturnValue(Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      requestVerifyEmail.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLoginMFA(mfaPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).toBeCalled();
      expect(loadSecurities).toBeCalled();
      expect(requestVerifyEmail).toBeCalled();
    });

    it('creates the correct actions with the correct payloads, when email verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = true;
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: stateBeforeAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.confirmSignIn.mockReturnValue(Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      requestVerifyEmail.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLoginMFA(mfaPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).toBeCalled();
      expect(loadSecurities).toBeCalled();
      expect(requestVerifyEmail).toBeCalled();
    });

    it('creates the correct actions with the correct payloads, with email and email additional not needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: stateBeforeAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN_MFA,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.confirmSignIn.mockReturnValue(Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLoginMFA(mfaPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.confirmSignIn).toBeCalledWith(userObject.user, mfaPayload.code);
      expect(loadAWSFields).toBeCalled();
      expect(loadSecurities).toBeCalled();
      expect(requestVerifyEmail).not.toBeCalled();
    });
  });
});
