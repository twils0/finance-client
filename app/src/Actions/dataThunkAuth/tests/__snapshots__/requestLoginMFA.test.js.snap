// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataThunkAuth requestLoginMFA fails and throws an error when missing 'code' key in payload 1`] = `[Error: Please enter a value for the 'code' key - {}]`;
